### DISCUSSION TITLE
Bulk inserts

### DISCUSSION THREAD

**Discussion by tetotechy** [geldata/gel#4326] [General]
Not a db superexpert here, but I would like to check whether this behavior is consistent. I have an empy set of `Country`s, a type with two props: `code` (.e.g, `ENG`) and `name` (e.g., `England`). There is an exclusivity `constraint` on `code`. Doing a bulk insert with the `unless conflict` clause, if there is any duplicate in the source data I get a `ConstraintViolationError`. So:
```
 select (insert Country { code := 'ENG', name := 'England' } unless conflict) union (insert Country { code := 'ENG', name := 'England' } unless conflict);
edgedb error: ConstraintViolationError: code violates exclusivity constraint
  Detail: value of property 'code' of object type 'default::Country' violates exclusivity constraint`
```
But if the record is already present in the type data, the same bulk insert works:
```
insert Country { code := 'ENG', name := 'England' } unless conflict;
{default::Country {id: b99b91dc-2e38-11ed-98ce-b782cf9ff3c7}}       // DATA INSERTED
```
and:
```
select (insert Country { code := 'ENG', name := 'England' } unless conflict) union (insert Country { code := 'ENG', name := 'England' } unless conflict);
{}       // DATA CORRECTLY NOT INSERTED BUT NO ERROR!
```
I wonder whether this is consistent. I would expect either a 'smart insertion' or an error in both cases, not a different behavior. Am I missing something? Thanks!

**Comment by msullivan** [comment]
Yeah, this is intended (and documented) behavior, though we may be able to fix it at some point.

In the `insert` documentation, we note:
> A caveat, however, is that unless conflict will not prevent conflicts caused between multiple DML operations in the same query; inserting two conflicting objects (through use of for or simply with two insert statements) will cause a constraint error.

The implementation issues involved in this are discussed in detail in https://github.com/edgedb/edgedb/issues/2193.

**Comment by tetotechy** [comment]
Thanks! for the quick answer and for pointing me in the right direction. #2193 is what I needed.
The only thing I can say, If you decide to stay with ERROR, maybe the docs might specify that the transaction won't ALWAYS raise an error -- It took me a while before crossing it, and to understand what was going on, whereas the same query was working smoothly before
